src/promise-timeout.js

```diff
@@ -0,0 +1,48 @@
+'use strict'
+
+class PromiseTimeoutError extends Error {
+  constructor (message = null) {
+    super(message)
+
+    if (!message) {
+      this.message = 'ERR_PROMISE_TIMEOUT'
+    }
+
+    this.name = this.constructor.name
+    Error.captureStackTrace(this, this.constructor)
+  }
+}
+
+/**
+ * Run a promise with an execution timeout
+ *
+ * @param {Promise<any>|Function} promise Promise that timeout will be checked against
+ * @param {number} ms Time in milliseconds until promise is timed out
+ * @param {string} [errMsg] Custom error message for PromiseTimeoutError
+ *
+ * @returns {Promise<any>}
+ * @throws {PromiseTimeoutError}
+ */
+const promiseTimeout = async (promise, ms, errMsg = null) => {
+  const _promise = typeof promise === 'function' ? promise() : promise
+  let timeout = null
+
+  const res = await Promise.race([
+    _promise,
+    new Promise(resolve => {
+      timeout = setTimeout(() => {
+        resolve(new PromiseTimeoutError(errMsg))
+      }, ms)
+    })
+  ])
+
+  if (res instanceof PromiseTimeoutError) throw res
```

Comment:
what happens with other exceptions? are they "eaten up"? can you add a test for the case?

Reply:
yep we already have test case for other errors: https://github.com/bitfinexcom/lib-js-util-promise/pull/1/files#diff-5cf8949559fe57d1e5466ca328999b67cc3f6e2ddc99d38fc3ec3c2454edc753R51

Reply:
hmm actually agree, looks more convenient with rejection